### PR TITLE
Strong context elements

### PR DESCRIPTION
**Pull request by pjanek**

### CODE REVIEW

File: lib/codequest_pipes/pipe.rb [hunk 1]
```diff
@@ -26,12 +26,18 @@ def self.call(ctx)
...
+      _required_context_elements.merge!(
+        **args.map { |a| [a, nil] }.to_h,
+        **kwargs
+      )
     end
 
-    def self.provide_context(*args)
-      _provided_context_elements.push(*args)
+    def self.provide_context(*args, **kwargs)
+      _provided_context_elements.merge!(
```
Comment (Xypus) [inline lib/codequest_pipes/pipe.rb]: Operations performed on `provide` and `require` arguments are basically the same. How about extracting those operations to one method?

File: lib/codequest_pipes/pipe.rb [hunk 2]
```diff
@@ -50,23 +56,36 @@ def self._check_interface(klass)
...
-      @provided_context_elements ||= []
+      @provided_context_elements ||= {}
     end
     private_class_method :_provided_context_elements
 
     def self._validate_ctx(collection, ctx)
-      collection.each do |element|
-        next if ctx.respond_to?(element)
-        fail MissingContext, "context does not respond to '#{element}'"
+      collection.each do |element, cls|
```
Comment (Xypus) [inline lib/codequest_pipes/pipe.rb]: If `cls` is supposed to be short for `Class`, I believe convention has a handy `klass` word for it.

File: lib/codequest_pipes/pipe.rb [hunk 3]
```diff
@@ -50,23 +56,36 @@ def self._check_interface(klass)
...
+        _raise_missing_context(element) unless ctx.respond_to?(element)
+        next unless cls
+        obj = ctx.public_send(element)
```
Comment (Xypus) [inline lib/codequest_pipes/pipe.rb]: I'd probably split this method into two: `validate_context_values_presence` and `validate_context_values_type` or sth.

File: spec/pipe_spec.rb [hunk 4]
```diff
@@ -82,6 +82,40 @@ class NotProvidingChild < Parent
...
+          add(bacon: true, eggs: 4)
+        end
+      end # class ProvideChild
+
+      let(:pipe) { Parent | ProvidingValidChild }
+
+      it 'does not raise' do
+        expect { subject }.to_not raise_error
+      end
+    end # context 'when context element with valid type provided'
```
Comment (Xypus) [inline spec/pipe_spec.rb]: I'd like to see some specs for `require_context` as well, even though they share much of the implementation.

**Review by Xypus — CHANGES_REQUESTED** [review]
Some cosmetic suggestions, but a missing spec as well.

File: lib/codequest_pipes/pipe.rb (hunk 1 above)
Comment (pjanek) [reply lib/codequest_pipes/pipe.rb]: 👍

File: lib/codequest_pipes/pipe.rb (hunk 2 above)
Comment (pjanek) [reply lib/codequest_pipes/pipe.rb]: 👍

File: spec/pipe_spec.rb (hunk 4 above)
Comment (pjanek) [reply spec/pipe_spec.rb]: 👍

File: lib/codequest_pipes/pipe.rb (hunk 3 above)
Comment (pjanek) [reply lib/codequest_pipes/pipe.rb]: I don't like the idea of iterating over `collection` two times, but I've done small refactoring here.

**Review by Xypus — APPROVED** [review]
🍪

**Review by ghost — APPROVED** [review]

**Review by michalkorzawski — APPROVED** [review]

**Pull request merged**